File: 백준/Silver/1874. 스택 수열/스택 수열.cc
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <stack>
#include <queue>
using namespace std;

int n;
queue<int> inputs;
stack<int> numbers;
queue<char> order;

int main() {
	scanf("%d", &n);

	//입력받기
	for (int i = 0; i < n; i++) {
		int tmp = 0;
		scanf("%d", &tmp);
		inputs.push(tmp);
	}

	int index = 1;
	while (inputs.empty() != true) {
		if (numbers.empty() != true && numbers.top() == inputs.front()) {
			numbers.pop();
			inputs.pop();
			order.push('-');
		}
		else if (index <= inputs.front()) {
			while (index <= inputs.front()) {
				numbers.push(index++);
				order.push('+');
			}
		}
		else {
			printf("NO");
			return 0;
		}
	}

	while (order.empty() != true) {
		printf("%c\n", order.front());
		order.pop();
	}
}

File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
from collections import deque

input = sys.stdin.readline

queue = deque()
stack = []

n = int(input())

for i in range(n):
    tmp = int(input())
    queue.append(tmp)

index = 1
answer = []
flag = 0

while queue:
    if stack and stack[-1] == queue[0]:
        stack.pop()
        queue.popleft()
        answer.append('-')
    elif index <= queue[0]:
        while index <= queue[0]:
            stack.append(index)
            index+=1
            answer.append('+')
    else:
        flag = -1
        break

if flag==-1:
    print('NO')
else:
    for a in answer:
        print(a, end=' ')
